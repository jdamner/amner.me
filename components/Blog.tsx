import React from "react";
/* Layout */
import PostList from "./PostList";
/* Types */
import ContactDetails from "./ContactDetails";

export default function Blog({
  posts,
}: {
  posts: { slug: string; title: string }[];
}) {
  return (
    <div className="lg:grid lg:grid-cols-3">
      <PostList posts={posts} className="mt-5 col-span-1" />
      <ContactDetails />
    </div>
  );
}
