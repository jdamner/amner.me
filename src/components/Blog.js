import React from "react";
import Title from "./Global/Title";

import PostList from "./PostList/PostList";


export default function Blog( { posts }) {
  return (
    <section className='blog' id='blog'>
      <div className='blog-wrap'>
        <Title text='My Projects and Writing' />
        <div className='blog-wrap-grid'>
          <PostList posts={posts} />
        </div>
        
      </div>
    </section>
  );
}