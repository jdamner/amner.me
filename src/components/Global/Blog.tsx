import React from "react";
/* Layout */
import Title from "./Title";
import PostList from "./PostList";
/* Types */
import type { PostLinkType } from "../../types";

export default function Blog({ posts }: { posts: PostLinkType[] }) {
  return (
    <>
      <Title title="Projects and Writing">Blog</Title>
      <PostList posts={posts} />
    </>
  );
}
