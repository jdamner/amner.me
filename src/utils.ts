/**
 * Helper functions
 */
import fs from 'fs'
import matter from 'gray-matter'
import { join } from 'path'
import { slug } from 'github-slugger'

import type { ReactNode } from 'react'
import type { MdFile } from './types'

/**
 * Makes an object JSON parseable
 * 
 * @param {any} obj The object to be parsed
 * @returns {any}
 */
export function makeJsonParseable<Type>(obj: Type): Type {
	return JSON.parse(JSON.stringify(obj)) as Type
}

/**
 * Reads a markdown file
 */
export async function readMdFile(filename: string, directory: string = 'content/'): Promise<MdFile> {
	const path = `${directory}/${filename}`
	const fullPath = `${process.cwd()}/${path}`
	const fileContents = fs.readFileSync(fullPath, 'utf8')
	return {
		slug: slug(filename.replace(/\.md$/, '')),
		...matter(fileContents),
	}
}

/**
 * Reads all slugs from a directory
 * 
 * @param {string} directory The directory to read from
 * 
 * @returns {string[]}
 */
export function getFilesInDir(directory: string): string[] {
	const fullPath = join(process.cwd(), directory)
	return fs.readdirSync(fullPath)
}

/**
 * Converts all headings in a markdown file to have IDs 
 * Converts 
 * ## Heading 1 
 * to become
 * ## Heading 1{#heading-1}
 * 
 * @param {string} markdown The markdown to convert
 * 
 * @returns {string}
 */
export function convertHeadingsToAnchorLinks(markdown: string): string {
	return markdown.replace(/(#{1,6})\s(.+)\n/gm, (match, p1, p2) => {
		return `${p1} ${p2}{#${slug(p2)}}\n`
	})
}


/**
 * Converts a react node to a slug
 * 
 * Usually the value is a string, but sometimes it contains
 * a `<code>` element, which is a react node. This function
 * converts the value to a string.
 */
export function makeSlug(value: ReactNode | ReactNode[]): string {
	if (! Array.isArray(value)) return slug(value?.toString() ?? '');
	return slug(
		value.map((node) => {
			if (typeof node === 'string' ||
				typeof node === 'number' ||
				typeof node === 'boolean'
			) return node;

			if (typeof node === 'object' && node !== null && 'props' in node) {
				return makeSlug(node.props.children)
			}

			return ''
		}).join('')
	)
}