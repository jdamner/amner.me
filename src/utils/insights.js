import { track, parameters, init } from 'insights-js';

const pageView = () => {
    event('page-view');
}

const event = (eventName, eventProperties = {}) => {
    init('HbSIygPWVvMzXFRP');
    track({
        id: eventName,
        parameters: {
            ...parameters,
            ...eventProperties,
        },
    });
}

export { pageView, event };