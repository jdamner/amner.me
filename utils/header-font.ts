import { Bebas_Neue as Font } from "next/font/google";
export const font = Font({
  weight: "400",
  subsets: ["latin", "latin-ext"],
});
