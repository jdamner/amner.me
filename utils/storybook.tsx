import { faker } from "@faker-js/faker";
import { MdFile, PostLinkType } from "../types";
import matter from "gray-matter";
import React from "react";

type MdAttributes = Record<string, string>;

faker.seed(123);

export const generateMdFile = <T extends MdAttributes>(
  attributes: T,
  content?: string,
): Omit<MdFile<string>, "data"> & { data: T } => {
  return matter(generateMD(attributes, content)) as Omit<
    MdFile<string>,
    "data"
  > & { data: T };
};

export const generateMdContent = (): string => {
  return (
    `# ${faker.lorem.words(3)}\n` +
    Array.from({ length: 3 }, () => {
      return (
        `## ${faker.lorem.words(2)}\n` +
        faker.lorem.paragraph() +
        "\n" +
        Array.from({ length: 3 }, () => {
          return (
            `### ${faker.lorem.words(2)}\n` + `${faker.lorem.paragraph()}\n`
          );
        }).join("")
      );
    }).join("")
  );
};

export const generatePostLink = (): PostLinkType => {
  return {
    title: faker.lorem.words(3),
    thumbnail: faker.image.url(),
    slug: faker.lorem.slug(),
  };
};

export const generatePostLinks = (count: number): PostLinkType[] => {
  return Array.from({ length: count }, () => generatePostLink());
};

export const generateMD = (
  attributes: MdAttributes,
  content?: string,
): string => {
  const frontMatter = generateMDFrontMatter(attributes);
  return `${frontMatter}\n${content ?? generateMdContent()}`;
};

export const generateMDFrontMatter = (attributes: MdAttributes): string => {
  return (
    "---\n" +
    Object.keys(attributes)
      .map((key) => `${key}: ${attributes[key]}`)
      .join("\n") +
    "\n---"
  );
};

export const generatePlaceholderContent = (
  size: number = 5,
): React.ReactNode => {
  return (
    <>
      {Array.from({ length: size }).map((_, i) => (
        <p key={i}>{faker.lorem.paragraph(i + 1)}</p>
      ))}
    </>
  );
};
