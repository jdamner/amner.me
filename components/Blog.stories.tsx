import type { Meta, StoryObj } from "@storybook/nextjs";

import Blog from "./Blog";
import { generatePostLinks } from "../utils/storybook";

type Story = StoryObj<typeof Blog>;

export default {
  title: "Components/Blog",
  component: Blog,
} as Meta;

export const Default: Story = {
  args: {
    posts: generatePostLinks(10),
  },
};
