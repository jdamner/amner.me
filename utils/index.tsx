/* External Deps */
import { readFileSync, readdirSync } from "fs";
import { resolve } from "path";
import matter from "gray-matter";
import { slug } from "github-slugger";

/* Components */
import { Markdown } from "@/components/Markdown";

/* Constants */
const postsDirectory: string = resolve("content/blog/");
const servicesDirectory: string = resolve("content/services");
const cvDirectory: string = resolve("content/cv");
const projectsDirectory: string = resolve(cvDirectory, "projects");
const referencesDirectory: string = resolve(cvDirectory, "references");
const employmentDirectory: string = resolve(cvDirectory, "employment");
const educationDirectory: string = resolve(cvDirectory, "education");

type MatterData = ReturnType<typeof matter>["data"];
type AttributeValidator<T extends MatterData> = (data: MatterData) => data is T;

const defaultAttributeValidator = <T extends MatterData>(
  data: MatterData,
): data is T => {
  return typeof data === "object" && data !== null;
};

/**
 * Gets all files in a directory
 */
const getAllFrom = <T extends MatterData>(
  directory: string,
  attributeValidator: AttributeValidator<T> = defaultAttributeValidator,
) =>
  readdirSync(directory).map((filename) =>
    readMdFile<T>(filename, directory, attributeValidator),
  );

/**
 * Reads a markdown file
 */
const readMdFile = <T extends MatterData>(
  filename: string,
  directory: string = "content/",
  attributeValidator: AttributeValidator<T> = defaultAttributeValidator,
) =>
  parseMarkdown<T>(
    readFileSync(`${directory}/${filename}`, "utf8"),
    filename,
    attributeValidator,
  );

/**
 * Parses markdown file contents
 */
export const parseMarkdown = <T extends MatterData>(
  fileContents: string,
  filename: string,
  attributeValidator: AttributeValidator<T> = defaultAttributeValidator,
) => {
  const { content, data } = matter(fileContents);

  if (!attributeValidator(data)) {
    console.log(data);
    throw new Error(`Invalid attributes in ${filename}` + JSON.stringify(data));
  }

  return {
    data,
    slug: slug(filename.replace(/\.mdx$/, "")),
    content: <Markdown>{content}</Markdown>,
  };
};

/**
 * Gets all Posts
 */
const getAllPosts = () =>
  getAllFrom<WithTitle & WithThumbnail & WithDate>(
    postsDirectory,
    (data): data is WithTitle & WithThumbnail & WithDate =>
      hasTitle(data) && hasThumbnail(data) && hasDate(data),
  );

export const useAllServices = () =>
  getAllFrom<WithTitle>(servicesDirectory, hasTitle);

export const getAllProjects = () =>
  getAllFrom<WithTitle & WithLink>(
    projectsDirectory,
    (data): data is WithTitle & WithLink => hasTitle(data) && hasLink(data),
  );

export const useAllReferences = () =>
  getAllFrom<WithTitle & WithWebsite & WithPhone & WithEmail>(
    referencesDirectory,
    (data): data is WithTitle & WithWebsite & WithPhone & WithEmail =>
      hasTitle(data) && hasWebsite(data) && hasPhone(data) && hasEmail(data),
  );

export const useAllEmployment = () =>
  getAllFrom<
    WithTitle & WithSubtitle & WithDateString & WithRole & WithWebsite
  >(
    employmentDirectory,
    (
      data,
    ): data is WithTitle &
      WithSubtitle &
      WithDateString &
      WithRole &
      WithWebsite =>
      hasTitle(data) &&
      hasSubtitle(data) &&
      hasDateString(data) &&
      hasRole(data) &&
      hasWebsite(data),
  );

export const useAllEducation = () =>
  getAllFrom<WithTitle & WithDateString>(
    educationDirectory,
    (data): data is WithTitle & WithDateString =>
      hasTitle(data) && hasDateString(data),
  );

export const getPost = (slug: string) =>
  getAllPosts().find((post) => post.slug === slug);

export const getAllPostLinks = () =>
  getAllPosts()
    .map((post) => ({
      title: post.data.title,
      thumbnail: post.data.thumbnail,
      slug: post.slug,
      date: post.data.date,
    }))
    .sort((a, b) => (a.date > b.date ? -1 : 1));

type WithTitle = { title: string };
type WithSubtitle = { subtitle: string };
type WithLink = { link: string };
type WithThumbnail = { thumbnail: string };
type WithDate = { date: Date };
type WithDateString = { date: string };
type WithPhone = { phone: string };
type WithWebsite = { website: string };
type WithEmail = { email: string };
type WithRole = { role: string };

const hasTitle = (data: MatterData): data is WithTitle =>
  defaultAttributeValidator(data) &&
  "title" in data &&
  typeof data.title === "string";

const hasSubtitle = (data: MatterData): data is WithSubtitle =>
  "subtitle" in data && typeof data.subtitle === "string";

const hasLink = (data: MatterData): data is WithLink =>
  "link" in data && typeof data.link === "string";

const hasThumbnail = (data: MatterData): data is WithThumbnail =>
  "thumbnail" in data && typeof data.thumbnail === "string";

const hasDate = (data: MatterData): data is WithDate =>
  "date" in data && data.date instanceof Date;

const hasDateString = (data: MatterData): data is WithDateString =>
  "date" in data && typeof data.date === "string";

const hasPhone = (data: MatterData): data is WithPhone =>
  "phone" in data && typeof data.phone === "string";

const hasWebsite = (data: MatterData): data is WithWebsite =>
  "website" in data && typeof data.website === "string";

const hasEmail = (data: MatterData): data is WithEmail =>
  "email" in data && typeof data.email === "string";

export const hasRole = (data: MatterData): data is WithRole =>
  "role" in data && typeof data.role === "string";
