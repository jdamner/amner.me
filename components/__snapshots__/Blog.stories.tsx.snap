// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Blog Default smoke-test 1`] = `
<div class="bg-slate-100 dark:bg-slate-900 text-slate-900 dark:text-slate-300">
  <div class="flex flex-col md:flex-row mb-5 gap-5 md:items-end">
    <div class="basis-1/3">
      <h2 class="text-lg font-sans font-semibold tracking-tight text-left font-slate-800 uppercase text-5xl md:text-6xl break-words basis-1/3">
        Projects and Writing
      </h2>
    </div>
    <div class="w-full">
      <div class="w-full text-left  text-2xl font-black">
        Blog
        <div class="bg-slate-800 dark:bg-slate-300 h-1 w-full"
             style="max-width: 100%;"
        >
        </div>
      </div>
    </div>
  </div>
  <div class="columns-1 xs:columns-2 md:columns-3 lg:columns-4 gap-2 space-y-2"
       id="post-list"
  >
    <div class="break-inside-avoid-column">
      <a class="group block bg-white dark:bg-slate-900 border-2 border-slate-500"
         href="/studio-ipsa-super"
      >
        <img alt
             loading="lazy"
             width="375"
             height="375"
             decoding="async"
             data-nimg="1"
             class="w-full brightness-90 dark:brightness:75 group-hover:brightness-105"
             srcset="https://loremflickr.com/640/480?lock=2043296255836160&amp;w=384&amp;q=75 1x, https://loremflickr.com/640/480?lock=2043296255836160&amp;w=750&amp;q=75 2x"
             src="https://loremflickr.com/640/480?lock=2043296255836160&amp;w=750&amp;q=75"
             style="color: transparent;"
        >
        <h2 class="bg-slate-900 group-hover:bg-slate-700 px-2 text-lg leading-loose text-center text-slate-100">
          subito summisse circumvenio
        </h2>
      </a>
    </div>
    <div class="break-inside-avoid-column">
      <a class="group block bg-white dark:bg-slate-900 border-2 border-slate-500"
         href="/commodo-decens-caries"
      >
        <img alt
             loading="lazy"
             width="375"
             height="375"
             decoding="async"
             data-nimg="1"
             class="w-full brightness-90 dark:brightness:75 group-hover:brightness-105"
             srcset="https://picsum.photos/seed/pGzt8oolCJ/640/480?w=384&amp;q=75 1x, https://picsum.photos/seed/pGzt8oolCJ/640/480?w=750&amp;q=75 2x"
             src="https://picsum.photos/seed/pGzt8oolCJ/640/480?w=750&amp;q=75"
             style="color: transparent;"
        >
        <h2 class="bg-slate-900 group-hover:bg-slate-700 px-2 text-lg leading-loose text-center text-slate-100">
          super deprecator curo
        </h2>
      </a>
    </div>
    <div class="break-inside-avoid-column">
      <a class="group block bg-white dark:bg-slate-900 border-2 border-slate-500"
         href="/reiciendis-coadunatio-ulciscor"
      >
        <img alt
             loading="lazy"
             width="375"
             height="375"
             decoding="async"
             data-nimg="1"
             class="w-full brightness-90 dark:brightness:75 group-hover:brightness-105"
             srcset="https://picsum.photos/seed/Dbra5wIwq/640/480?w=384&amp;q=75 1x, https://picsum.photos/seed/Dbra5wIwq/640/480?w=750&amp;q=75 2x"
             src="https://picsum.photos/seed/Dbra5wIwq/640/480?w=750&amp;q=75"
             style="color: transparent;"
        >
        <h2 class="bg-slate-900 group-hover:bg-slate-700 px-2 text-lg leading-loose text-center text-slate-100">
          adulatio sub cubicularis
        </h2>
      </a>
    </div>
    <div class="break-inside-avoid-column">
      <a class="group block bg-white dark:bg-slate-900 border-2 border-slate-500"
         href="/canonicus-claudeo-voveo"
      >
        <img alt
             loading="lazy"
             width="375"
             height="375"
             decoding="async"
             data-nimg="1"
             class="w-full brightness-90 dark:brightness:75 group-hover:brightness-105"
             srcset="https://picsum.photos/seed/IpkEmne/640/480?w=384&amp;q=75 1x, https://picsum.photos/seed/IpkEmne/640/480?w=750&amp;q=75 2x"
             src="https://picsum.photos/seed/IpkEmne/640/480?w=750&amp;q=75"
             style="color: transparent;"
        >
        <h2 class="bg-slate-900 group-hover:bg-slate-700 px-2 text-lg leading-loose text-center text-slate-100">
          deprimo surculus taedium
        </h2>
      </a>
    </div>
    <div class="break-inside-avoid-column">
      <a class="group block bg-white dark:bg-slate-900 border-2 border-slate-500"
         href="/color-valetudo-cursim"
      >
        <img alt
             loading="lazy"
             width="375"
             height="375"
             decoding="async"
             data-nimg="1"
             class="w-full brightness-90 dark:brightness:75 group-hover:brightness-105"
             srcset="https://picsum.photos/seed/1qau9q4/640/480?w=384&amp;q=75 1x, https://picsum.photos/seed/1qau9q4/640/480?w=750&amp;q=75 2x"
             src="https://picsum.photos/seed/1qau9q4/640/480?w=750&amp;q=75"
             style="color: transparent;"
        >
        <h2 class="bg-slate-900 group-hover:bg-slate-700 px-2 text-lg leading-loose text-center text-slate-100">
          quos tener aliquam
        </h2>
      </a>
    </div>
    <div class="break-inside-avoid-column">
      <a class="group block bg-white dark:bg-slate-900 border-2 border-slate-500"
         href="/cuppedia-statua-usitas"
      >
        <img alt
             loading="lazy"
             width="375"
             height="375"
             decoding="async"
             data-nimg="1"
             class="w-full brightness-90 dark:brightness:75 group-hover:brightness-105"
             srcset="https://picsum.photos/seed/vsCz7iju/640/480?w=384&amp;q=75 1x, https://picsum.photos/seed/vsCz7iju/640/480?w=750&amp;q=75 2x"
             src="https://picsum.photos/seed/vsCz7iju/640/480?w=750&amp;q=75"
             style="color: transparent;"
        >
        <h2 class="bg-slate-900 group-hover:bg-slate-700 px-2 text-lg leading-loose text-center text-slate-100">
          tandem varietas suadeo
        </h2>
      </a>
    </div>
    <div class="break-inside-avoid-column">
      <a class="group block bg-white dark:bg-slate-900 border-2 border-slate-500"
         href="/voro-candidus-earum"
      >
        <img alt
             loading="lazy"
             width="375"
             height="375"
             decoding="async"
             data-nimg="1"
             class="w-full brightness-90 dark:brightness:75 group-hover:brightness-105"
             srcset="https://loremflickr.com/640/480?lock=4462993249140736&amp;w=384&amp;q=75 1x, https://loremflickr.com/640/480?lock=4462993249140736&amp;w=750&amp;q=75 2x"
             src="https://loremflickr.com/640/480?lock=4462993249140736&amp;w=750&amp;q=75"
             style="color: transparent;"
        >
        <h2 class="bg-slate-900 group-hover:bg-slate-700 px-2 text-lg leading-loose text-center text-slate-100">
          veritatis casso amitto
        </h2>
      </a>
    </div>
    <div class="break-inside-avoid-column">
      <a class="group block bg-white dark:bg-slate-900 border-2 border-slate-500"
         href="/trepide-thema-pax"
      >
        <img alt
             loading="lazy"
             width="375"
             height="375"
             decoding="async"
             data-nimg="1"
             class="w-full brightness-90 dark:brightness:75 group-hover:brightness-105"
             srcset="https://loremflickr.com/640/480?lock=4331491186180096&amp;w=384&amp;q=75 1x, https://loremflickr.com/640/480?lock=4331491186180096&amp;w=750&amp;q=75 2x"
             src="https://loremflickr.com/640/480?lock=4331491186180096&amp;w=750&amp;q=75"
             style="color: transparent;"
        >
        <h2 class="bg-slate-900 group-hover:bg-slate-700 px-2 text-lg leading-loose text-center text-slate-100">
          debilito placeat tantillus
        </h2>
      </a>
    </div>
    <div class="break-inside-avoid-column">
      <a class="group block bg-white dark:bg-slate-900 border-2 border-slate-500"
         href="/cognomen-confugo-cur"
      >
        <img alt
             loading="lazy"
             width="375"
             height="375"
             decoding="async"
             data-nimg="1"
             class="w-full brightness-90 dark:brightness:75 group-hover:brightness-105"
             srcset="https://loremflickr.com/640/480?lock=87886405828608&amp;w=384&amp;q=75 1x, https://loremflickr.com/640/480?lock=87886405828608&amp;w=750&amp;q=75 2x"
             src="https://loremflickr.com/640/480?lock=87886405828608&amp;w=750&amp;q=75"
             style="color: transparent;"
        >
        <h2 class="bg-slate-900 group-hover:bg-slate-700 px-2 text-lg leading-loose text-center text-slate-100">
          cimentarius impedit damnatio
        </h2>
      </a>
    </div>
    <div class="break-inside-avoid-column">
      <a class="group block bg-white dark:bg-slate-900 border-2 border-slate-500"
         href="/quibusdam-cerno-spero"
      >
        <img alt
             loading="lazy"
             width="375"
             height="375"
             decoding="async"
             data-nimg="1"
             class="w-full brightness-90 dark:brightness:75 group-hover:brightness-105"
             srcset="https://picsum.photos/seed/vsFgAP/640/480?w=384&amp;q=75 1x, https://picsum.photos/seed/vsFgAP/640/480?w=750&amp;q=75 2x"
             src="https://picsum.photos/seed/vsFgAP/640/480?w=750&amp;q=75"
             style="color: transparent;"
        >
        <h2 class="bg-slate-900 group-hover:bg-slate-700 px-2 text-lg leading-loose text-center text-slate-100">
          strenuus stabilis ustulo
        </h2>
      </a>
    </div>
  </div>
</div>
`;
